log4j-redis/src/main/java/org/apache/logging/log4j/redis/appender/RedisAppender.java

```diff
@@ -0,0 +1,244 @@
+            return poolConfiguration;
+        }
+
+        int getPort() {
+            return port;
+        }
+
+        public B setKeys(final String keys) {
+            this.keys = keys;
+            return asBuilder();
+        }
+
+        public B setHost(final String host) {
+            this.host = host;
+            return asBuilder();
+        }
+
+        public B setPort(final int port) {
+            this.port = port;
+            return asBuilder();
+        }
+
+        public B setQueueCapacity(final int queueCapacity) {
+            this.queueCapacity = queueCapacity;
+            return asBuilder();
+        }
+
+        public B setPoolConfiguration(final LoggingJedisPoolConfiguration poolConfiguration) {
+            this.poolConfiguration = poolConfiguration;
+            return asBuilder();
+        }
+
+        public B setSslConfiguration(final SslConfiguration ssl) {
+            this.sslConfiguration = ssl;
+            return asBuilder();
+        }
+
+        public B setImmediateFlush(final boolean immediateFlush) {
+            this.immediateFlush = immediateFlush;
+            return asBuilder();
+        }
+
+        RedisManager getRedisManager() {
+            return new RedisManager(
+                    getConfiguration().getLoggerContext(),
+                    getName(),
+                    getKeys().split(KEY_SEPARATOR),
+                    getHost(),
+                    getPort(),
+                    getSslConfiguration(),
+                    getPoolConfiguration()
+            );
+        }
+    }
+
+    /**
+     * Creates a builder for a RedisAppender.
+     * @return a builder for a RedisAppender.
+     */
+    @PluginBuilderFactory
+    public static <B extends Builder<B>> B newBuilder() {
+        return new Builder<B>().asBuilder();
+    }
+
+    @Override
+    public void append(final LogEvent event) {
+        final Layout<? extends Serializable> layout = getLayout();
+        if (event.getLoggerName() != null && AbstractLogger.getRecursionDepth() > 1) {
+            LOGGER.warn("Recursive logging from [{}] for appender [{}].", event.getLoggerName(), getName());
+        } else if (layout instanceof StringLayout) {
+            logQueue.add(((StringLayout)layout).toSerializable(event));
+            if (shouldFlushLogQueue(event.isEndOfBatch())) {
+                try {
+                    tryFlushQueue();
+                } catch (final Exception e) {
+                    error("Unable to write to Redis in appender [" + getName() + "]", event, e);
+                }
+            }
+        } else {
+            throw new AppenderLoggingException("The Redis appender only supports StringLayouts.");
+        }
+    }
+
+    private boolean shouldFlushLogQueue(boolean endOfBatch) {
+        return immediateFlush || endOfBatch || logQueue.size() >= queueCapacity;
+    }
+
+    private void tryFlushQueue() {
+        manager.sendBulk(logQueue);
```

Comment:
We may want to use [BlockingQueue.drainTo](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/BlockingQueue.html#drainTo(java.util.Collection)) in this call

Reply:
I had initially implemented it that way, but my thinking on using `.poll()` in the inner method is that we can abort on a single failed log and keep the queue intact as opposed to dumping to a set and having to refill the queue with those logs that didn't throw exceptions and haven't yet been sent.

Is there something I'm missing that would make drainTo preferred?

Reply:
The client should ideally use bulk operations to reduce overhead and increase throughput. We can use drainTo in order to get the bag of values for which to make bulk operations.

A load spike will produce events faster than we can persist them. Without bulk operations, many threads are likely to fill the queue and individually attempt to flush, forcing us to do even more work overall.

Reply:
Gotcha -- I'm on the same page there. I thought the jedis.rpush API didn't support that, but it looks like the second argument is `String...`. I'll use drainTo instead, though this could increase our risk of dropping logs more broadly because a failure could mean a whole queue loss (Redis is pretty forgiving in what you send it though).